193 modularize error handling into a separate module

hichri-louay opened the pull request:
## PR Description

Hey team,

This PR introduces a middleware for error handling in our Node.js application. Currently, our application lacks a centralized mechanism for handling errors, making it difficult to manage and respond to errors consistently. To address this, we have implemented a middleware that will handle errors in a centralized manner.

## Changes Made

- Created a new middleware for error handling in our Node.js application.
- The error handling middleware captures any errors that occur during the request/response lifecycle and provides a centralized mechanism for handling them.
- The middleware ensures that errors are properly logged and appropriate responses are sent back to the client, with consistent error messages and status codes.
- The error handling middleware is added to the application's middleware stack to ensure it is executed for every incoming request.


## Notes for Reviewers

- Kindly review the changes made to ensure that the error handling middleware effectively captures and handles errors in a centralized manner.
- Verify that the middleware is correctly integrated into the application's middleware stack.
- Pay attention to the logging of errors and the consistency of error messages and status codes.
- If there are alternative approaches or suggestions to improve the implementation, please feel free to provide feedback.

Thank you for your time and consideration. Your feedback and suggestions are highly appreciated.

Best regards,
Louay HICHRI

BrosG commented:
Your error handling functions are well-structured and cover the most common scenarios you'd need to handle in an Express.js application. However, there are a couple of improvements that could make the code even better:

Specific error handling: The errorHandler function treats all errors as equal. However, in practice, you might want to handle specific errors differently. For example, if a validation error occurs, you might want to send a 400 status code instead of 500. One way to do this is to check the error type or name and add conditional logic for different error types.

Error logging: It's often useful to log errors for debugging or auditing purposes. You could add code to log the errors to the console, a file, or a logging service.
Here's an updated version of the errorHandler function incorporating these improvements:

`const errorHandler = (err, req, res, next) => {
    // Log the error (could also be to a file or external service)
    console.error(err);

    if (err) {
        let status = err.status || 500;
        let message = err.message || 'An error occurred during the request.';
        let name = err.name || 'Error';

        // Handle specific error types
        if (err.name === 'ValidationError') {
            status = 400;
            message = err.message;
        }

        const body = {
            name, 
            message,
            status
        } 

        // render the error page
        res.status(status).json(body);
    } else {
        next();
    }
}
`

hichri-louay commented:
Thank you for reviewing the error handling functions in our Express.js
application. I appreciate your feedback and suggestions for improvement. I
wanted to provide an update on the changes made based on your
recommendations.
I have made the following modifications to the errorHandler function:
      - Added console.log statements to log errors specifically for the
testnet environment.
      - Removed the check for error type, allowing us to view all errors in
the logs.
These changes will allow us to log errors in the testnet environment for
better debugging and troubleshooting purposes. By removing the check for
error type, we can capture and review all errors that occur.

PR : https://github.com/SaTT-Wallet/Backend/pull/226


Le ven. 26 mai 2023 à 12:49, Bros G ***@***.***> a écrit :

> Your error handling functions are well-structured and cover the most
> common scenarios you'd need to handle in an Express.js application.
> However, there are a couple of improvements that could make the code even
> better:
>
> Specific error handling: The errorHandler function treats all errors as
> equal. However, in practice, you might want to handle specific errors
> differently. For example, if a validation error occurs, you might want to
> send a 400 status code instead of 500. One way to do this is to check the
> error type or name and add conditional logic for different error types.
>
> Error logging: It's often useful to log errors for debugging or auditing
> purposes. You could add code to log the errors to the console, a file, or a
> logging service.
> Here's an updated version of the errorHandler function incorporating these
> improvements:
>
> `const errorHandler = (err, req, res, next) => {
> // Log the error (could also be to a file or external service)
> console.error(err);
>
> if (err) {
>     let status = err.status || 500;
>     let message = err.message || 'An error occurred during the request.';
>     let name = err.name || 'Error';
>
>     // Handle specific error types
>     if (err.name === 'ValidationError') {
>         status = 400;
>         message = err.message;
>     }
>
>     const body = {
>         name,
>         message,
>         status
>     }
>
>     // render the error page
>     res.status(status).json(body);
> } else {
>     next();
> }
>
> }
> `
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SaTT-Wallet/Backend/pull/221#issuecomment-1564272374>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A5UYTAHG6J5TSLHBODMNF63XICKF7ANCNFSM6AAAAAAYQALZCM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
